lib/updatehub/updatehub.c

```diff
@@ -0,0 +1,807 @@
+	struct coap_block_context block;
+	struct k_sem semaphore;
+	struct flash_img_context flash_ctx;
+	struct tc_sha256_state_struct sha256sum;
+	enum updatehub_response code_status;
+	u8_t uri_path[MAX_PATH_SIZE];
+	u8_t payload[MAX_PAYLOAD_SIZE];
+	int downloaded_size;
+	struct pollfd fds[1];
+	int sock;
+	int nfds;
+} ctx;
+
+static struct update_info {
+	char package_uid[TC_SHA256_BLOCK_SIZE + 1];
+	char sha256sum_image[TC_SHA256_BLOCK_SIZE + 1];
+	int image_size;
+} update_info;
+
+static void wait_fds(void)
+{
+	if (poll(ctx.fds, ctx.nfds, NETWORK_TIMEOUT) < 0) {
+		LOG_ERR("Error in poll");
+	}
+}
+
+static void prepare_fds(void)
+{
+	ctx.fds[ctx.nfds].fd = ctx.sock;
+	ctx.fds[ctx.nfds].events = 1;
+	ctx.nfds++;
+}
+
+static int metadata_hash_get(char *metadata)
+{
+	struct tc_sha256_state_struct sha256sum;
+	unsigned char hash[TC_SHA256_DIGEST_SIZE];
+	char buffer[2];
+	int buffer_len = 0;
+
+	if (tc_sha256_init(&sha256sum) == 0) {
+		return -1;
+	}
+
+	if (tc_sha256_update(&sha256sum, metadata, strlen(metadata)) == 0) {
+		return -1;
+	}
+
+	if (tc_sha256_final(hash, &sha256sum) == 0) {
+		return -1;
+	}
+
+	memset(update_info.package_uid, 0, TC_SHA256_BLOCK_SIZE + 1);
+	for (int i = 0; i < TC_SHA256_DIGEST_SIZE; i++) {
+		snprintk(buffer, TC_SHA256_DIGEST_SIZE, "%02x",
+			 hash[i]);
+		buffer_len = buffer_len + strlen(buffer);
+		strncat(&update_info.package_uid[i], buffer,
+			MIN(TC_SHA256_BLOCK_SIZE, buffer_len));
+	}
+
+	return 0;
+}
+
+static bool
+is_compatible_hardware(struct resp_probe_some_boards *metadata_some_boards)
+{
+	int i;
+
+	for (i = 0; i < metadata_some_boards->supported_hardware_len; i++) {
+		if (strncmp(metadata_some_boards->supported_hardware[i],
+			    CONFIG_BOARD, strlen(CONFIG_BOARD)) == 0) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void dns_cb(enum dns_resolve_status status, struct dns_addrinfo *info,
+		   void *ip)
+{
+	struct sockaddr_in *addr;
+	char hr_addr[NET_IPV6_ADDR_LEN], *hr_family, *buffer;
+	int buffer_len = 0;
+
+	switch (status) {
+	case DNS_EAI_CANCELED:
+		LOG_ERR("DNS query was canceled");
+		break;
+	case DNS_EAI_FAIL:
+		LOG_ERR("DNS resolve failed");
+		break;
+	case DNS_EAI_NODATA:
+		LOG_ERR("Cannot resolve address");
+		break;
+	case DNS_EAI_ALLDONE:
+		break;
+	case DNS_EAI_INPROGRESS:
+		if (info && info->ai_family == AF_INET) {
+			hr_family = "IPv4";
+			addr = &net_sin(&info->ai_addr)->sin_addr;
+			buffer = net_addr_ntop(info->ai_family, addr, hr_addr, sizeof(hr_addr));
+			buffer_len = strlen(buffer);
+			memset(ip, 0, MAX_IP_SIZE);
+			memcpy(ip, buffer, MIN(MAX_IP_SIZE, buffer_len));
+		}
+
+		return;
+
+	default:
+		LOG_ERR("DNS resolving error");
+		break;
+	}
+
+	k_sem_give(&ctx.semaphore);
+}
+
+static bool start_coap_client(void)
+{
+	static struct sockaddr_in addr = { .sin_family = AF_INET,
+					   .sin_port = htons(5683) };
+	char ip[MAX_IP_SIZE] = "UNKNOW_IP";
+
+	if (dns_get_addr_info(UPDATEHUB_SERVER, DNS_QUERY_TYPE_A, NULL,
```

Comment:
This is purely a style comment, but there is a blocking DNS API in the socket layer:
https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/socket.h#L206
(Usable as `getaddrinfo()`)

Example usage in a sample: https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/big_http_download/src/big_http_download.c#L277

Seems you could simplify here and get future benefits by using that API.  @pfalcon can weigh in, but most other socket samples using DNS are using the socket-based API.